### DISCUSSION TITLE
How enable Debug Layer in D3D12?

### DISCUSSION THREAD

**Discussion by snpkhy** [amerkoleci/Vortice.Windows#449] [Q&A]
I treid this:
`#if DEBUG
        {
            ID3D12Debug? debugController;
            if (D3D12.D3D12GetDebugInterface(out debugController).Failure)
            {
                throw new Exception("Failed to enabled debug layer.");
            }
            debugController?.EnableDebugLayer();
        }
#endif`
But it is what it is, it does not have, in other words, it does nothing

**Comment by MarcioAB** [comment]
It sounds good to me. If the mixed-mode debugging is enabled and if there are DX messages, those messages should be appearing in the Debug Output window.

There are 2 ways to get Debug Layer messages: 

1. Directly in the Debug Output window (need mixed-mode debugging enabled)
2. Using GetMessage(..) function (NO need mixed-mode debugging enabled)

1) To get the Debug Layer messages directly in the Debug Output window it is necessary to enable mixed-mode debugging because Debug Layer messages are from non-managed code. To me, the disadvantage in this case is that these DX messages are mixed with the other messages in the Debug Output window. To get these messages just use EnableDebugLayer() function

   Vortice.Direct3D12.D3D12.D3D12GetDebugInterface(out ID3D12Debug3 debug);
   debug.EnableDebugLayer();
   debug.SetEnableGPUBasedValidation(false);
   debug.Dispose();

2) Let me know if there is interest in getting the Debug Layer messages programmatically, so I will not extend this post even more.
